Alliance.Client/Extensions/ExNativeUI/AgentStatus/ViewModels/AgentInteractionInterfaceVMCustom.cs

```diff
@@ -0,0 +1,616 @@
+				}
+			}
+		}
+
+		[DataSourceProperty]
+		public string BackgroundColor
+		{
+			get
+			{
+				return _backgroundColor;
+			}
+			set
+			{
+				if (_backgroundColor != value)
+				{
+					_backgroundColor = value;
+					OnPropertyChangedWithValue(value, "BackgroundColor");
+				}
+			}
+		}
+
+		[DataSourceProperty]
+		public string TextColor
+		{
+			get
+			{
+				return _textColor;
+			}
+			set
+			{
+				if (_textColor != value)
+				{
+					_textColor = value;
+					OnPropertyChangedWithValue(value, "TextColor");
+				}
+			}
+		}
+
+		[DataSourceProperty]
+		public bool IsActive
+		{
+			get
+			{
+				return _isActive;
+			}
+			set
+			{
+				if (value != _isActive)
+				{
+					_isActive = value;
+					OnPropertyChangedWithValue(value, "IsActive");
+					if (!value)
+					{
+						ShowHealthBar = false;
+						PrimaryInteractionMessage = "";
+						SecondaryInteractionMessage = "";
+					}
+				}
+			}
+		}
+
+		public AgentInteractionInterfaceVMCustom(Mission mission)
+		{
+			_mission = mission;
+			IsActive = false;
+		}
+
+		internal void Tick()
+		{
+			if (IsActive && _mission.Mode == MissionMode.StartUp && _currentFocusedObject is Agent && ((Agent)_currentFocusedObject).IsEnemyOf(_mission.MainAgent))
+			{
+				IsActive = false;
+			}
+		}
+
+		internal void CheckAndClearFocusedAgent(Agent agent)
+		{
+			if (_currentFocusedObject != null && _currentFocusedObject as Agent == agent)
+			{
+				IsActive = false;
+				ResetFocus();
+				SecondaryInteractionMessage = "";
+			}
+		}
+
+		public void OnFocusedHealthChanged(IFocusable focusable, float healthPercentage, bool hideHealthbarWhenFull)
+		{
+			SetHealth(healthPercentage, hideHealthbarWhenFull);
+		}
+
+		internal void OnFocusGained(Agent mainAgent, IFocusable focusableObject, bool isInteractable)
+		{
+			if (!IsPlayerActive || (_currentFocusedObject == focusableObject && _currentObjectInteractable == isInteractable))
+			{
+				return;
+			}
+
+			ResetFocus();
+			_currentFocusedObject = focusableObject;
+			_currentObjectInteractable = isInteractable;
+			if (focusableObject is Agent agent)
+			{
+				if (agent.IsHuman)
+				{
+					SetAgent(mainAgent, agent, isInteractable);
+				}
+				else
+				{
+					SetMount(mainAgent, agent, isInteractable);
+				}
+			}
+			else if (focusableObject is UsableMissionObject usableMissionObject)
+			{
+				if (usableMissionObject is SpawnedItemEntity spawnedItemEntity)
+				{
+					bool canQuickPickup = Agent.Main.CanQuickPickUp(spawnedItemEntity);
+					SetItem(spawnedItemEntity, canQuickPickup, isInteractable);
+				}
+				else
+				{
+					SetUsableMissionObject(usableMissionObject, isInteractable);
+				}
+			}
+			else if (focusableObject is UsableMachine machine)
+			{
+				SetUsableMachine(machine, isInteractable);
+			}
+			else if (focusableObject is DestructableComponent machine2)
+			{
+				SetDestructibleComponent(machine2, isInteractable: false);
+			}
+		}
```

Comment:
I would refactor like that in order to facilitate readability

```suggestion
                internal void OnFocusGained(Agent mainAgent, IFocusable focusableObject, bool isInteractable)
                {
                    if (!IsPlayerActive || (_currentFocusedObject == focusableObject && _currentObjectInteractable == isInteractable))
                    {
                        return;
                    }
                
                    ResetFocus();
                    _currentFocusedObject = focusableObject;
                    _currentObjectInteractable = isInteractable;
                
                    switch (focusableObject)
                    {
                        case Agent agent:
                            agent.IsHuman ?
                                SetAgent(mainAgent, agent, isInteractable):
                                SetMount(mainAgent, agent, isInteractable);
                            break;
                            
                        case SpawnedItemEntity spawnedItem:
                            bool canQuickPickup = Agent.Main.CanQuickPickUp(spawnedItem);
                            SetItem(spawnedItem, canQuickPickup, isInteractable);
                            break;
                            
                        case UsableMissionObject usableObject:
                            SetUsableMissionObject(usableObject, isInteractable);
                            break;
                            
                        case UsableMachine machine:
                            SetUsableMachine(machine, isInteractable);
                            break;
                            
                        case DestructableComponent destructable:
                            SetDestructibleComponent(destructable, isInteractable: false);
                            break;
                    }
                }
```

Reply:
This part is unchanged decompiled talewords code. Not the best but we stink to it if possible